crypto/bio/bio_lib.c

```diff
@@ -856,21 +863,21 @@ int BIO_connect_retry(BIO *bio, int timeout)
 
     if (rv <= 0) {
         if (get_last_sys_error() == ETIMEDOUT) {
-            /*
-             * if blocking, despite blocking BIO, BIO_do_handshake() timed out
-             * when non-blocking, BIO_do_handshake() timed out early
-             * with rv == -1 and get_last_sys_error() == 0
+            /*-
+             * If blocking, BIO_do_handshake() timed out despite blocking BIO.
```

Comment:
This is probably a misunderstanding how a blocking connect works.
If the server or a firewall does not send RST packets the connect
will time-out after 1-2 minutes.  So a connection error may be
immediately due to a RST packet, or after a timeout due to no SYN-ACK packet.
In both cases the next address in this list should be tried,
I wonder if the BIO_reset resets that address.

Reply:
Thanks for that explanation.
AFAICS from `bio_conn.c` BIO_reset() resets the list of addresses to try, so does too much, right?

Reply:
Could be possible

Reply:
So I've just added for now the following comment:
```
            /*
             * TODO BIO_reset() maybe needs to be replaced since it likely
             * does too much, reseting the list of IP addresses to try
             */
```

Reply:
I've meanwhile taken the time to re-do my connection error experiments and made adaptations in this PR accordingly.
* For non-blocking connects the first call to `BIO_do_connect()` always returns with "BIO should retry" and `errno=115` (Operation now in progress) and obviously needs to be retried, which works fine.
* For local ports (e.g., 127.0.0.1 or localhost:8000) that no service listens to there is an immediate "Connection refused" error, which I'd say is fine and there is likely no need to retry connecting until timeout.
* If there is no route to a non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) there is an immediate "No route to host" error, which I'd say is fine and there is likely no need to retry connecting until timeout.
* For unreachable non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) an internal timeout occurs after 30 seconds and `get_last_socket_error()` or `BIO_sock_error()` yields `110` (Connection timed out), where a blocking connect sets `errno=110` while a non-blocking connect for some reason retains `errno=115. In both blocking and non-blocking case `BIO_connect_retry()` should retry (including retry of any applicable IP addresses).
If the retry is done directly I cannot reproduce a crash, but a connect error is thrown again immediately. Therefore `ERR_clear_error(); (void)BIO_reset(bio);` are really necessary.
For enabling `BIO_connect_retry()` to correctly detecting these internal timeout (`ETIMEDOUT`) situations I've added for both the blocking and non-blocking case a differentiation of the error reason thrown: `BIO_R_CONNECT_TIMEOUT` and `BIO_R_NBIO_CONNECT_TIMEOUT`, respectively.
* For DNS trouble on non-local addresses (e.g., 1.2.3.4:8000 or test.com:8000) an internal timeout occurs after 40 seconds and `BIOerr(BIO_F_BIO_LOOKUP_EX, ERR_R_SYS_LIB)` is thrown and `get_last_sys_error()` yields `EAGAIN` (resource temporarily unavailable) and `BIO_connect_retry()` should retry.

The adaptations lead to major cleanup of `BIO_connect_retry()`.

Reply:
> while a non-blocking connect for some reason retains errno=115.

I think that is basically what I claim that the errno is undefined when
the non-blocking socket connect fails:
```C
           i = BIO_sock_error(b->num);
            if (i != 0) {
                BIO_clear_retry_flags(b);
                if ((c->addr_iter = BIO_ADDRINFO_next(c->addr_iter)) != NULL) {
                    /*
                     * if there are more addresses to try, do that first
                     */
                    BIO_closesocket(b->num);
                    c->state = BIO_CONN_S_CREATE_SOCKET;
                    ERR_clear_error();
                    break;
                }
                ERR_raise_data(ERR_LIB_SYS, i,
                               "calling connect(%s, %s)",
                                c->param_hostname, c->param_service);
                BIOerr(BIO_F_CONN_STATE, BIO_R_NBIO_CONNECT_ERROR);
                ret = 0;
                goto exit_loop;
```
does nowere set errno, ETIMEDOUT
while `i = BIO_sock_error(b->num)` should return ETIMEDOUT,
so errno is still 115 EINPROGRESS.

Reply:
Yes, so I've meanwhile made sure that `BIO_do_retry()` is independent of `errno`,
which in my view is cleaner anyway.